geode-core/src/main/java/org/apache/geode/internal/net/BufferPool.java

```diff
@@ -82,123 +83,92 @@ private ByteBuffer acquireBuffer(int size, boolean send) {
     ByteBuffer result;
 
     if (useDirectBuffers) {
-      if (size <= Connection.SMALL_BUFFER_SIZE) {
-        return acquireSmallBuffer(send);
-      } else if (size <= DistributionConfig.DEFAULT_SOCKET_BUFFER_SIZE) {
-        return acquireMiddleBuffer(send);
-      }
-
-      IdentityHashMap<BBSoftReference, BBSoftReference> alreadySeen = null; // keys are used like a
-                                                                            // set
-      BBSoftReference ref = bufferQueue.poll();
-      while (ref != null) {
-        ByteBuffer bb = ref.getBB();
-        if (bb == null) {
-
-          // it was garbage collected
-          int refSize = ref.consumeSize();
-          if (refSize > 0) {
-            if (ref.getSend()) { // fix bug 46773
-              stats.incSenderBufferSize(-refSize, true);
-            } else {
-              stats.incReceiverBufferSize(-refSize, true);
-            }
-          }
-        } else if (bb.capacity() >= size) {
-
-          bb.rewind();
-          bb.limit(size);
-          return bb;
-        } else {
-
-          // wasn't big enough so put it back in the queue
-          Assert.assertTrue(bufferQueue.offer(ref));
-          if (alreadySeen == null) {
-            alreadySeen = new IdentityHashMap<>();
-          }
-          if (alreadySeen.put(ref, ref) != null) {
-            // if it returns non-null then we have already seen this item
-            // so we have worked all the way through the queue once.
-            // So it is time to give up and allocate a new buffer.
-            break;
-          }
-        }
-        ref = bufferQueue.poll();
+      if (size <= DistributionConfig.DEFAULT_SOCKET_BUFFER_SIZE) {
+        return acquirePredefinedFixedBuffer(send, size);
+      } else {
+        return acquireLargeBuffer(send, size);
       }
-      result = ByteBuffer.allocateDirect(size);
     } else {
       // if we are using heap buffers then don't bother with keeping them around
       result = ByteBuffer.allocate(size);
     }
-    if (send) {
-      stats.incSenderBufferSize(size, useDirectBuffers);
-    } else {
-      stats.incReceiverBufferSize(size, useDirectBuffers);
-    }
+    updateBufferStats(size, send, false);
     return result;
   }
 
-  private ByteBuffer acquireSmallBuffer(boolean send) {
+  /**
+   * Acquire direct buffer with predefined default capacity (4096 or 32768)
+   */
+  private ByteBuffer acquirePredefinedFixedBuffer(boolean send, int size) {
     // set
-    int size = Connection.SMALL_BUFFER_SIZE;
+    int defaultSize;
+    ConcurrentLinkedQueue<BBSoftReference> bufferTempQueue;
     ByteBuffer result;
 
-    BBSoftReference ref = bufferSmallQueue.poll();
+    if (size <= Connection.SMALL_BUFFER_SIZE) {
```

Comment:
Something that might make this code easier to understand would be to define two private constants in this class SMALL_BUFFER_SIZE and MEDIUM_BUFFER_SIZE. When you define them you can initialize them to Connection.SMALL_BUFFER_SIZE and DistributionConfig.DEFAULT_SOCKET_BUFFER_SIZE but I think having those better local constant names will help throughout this class as you do stuff with the smallQueue and middleQueue